### PR TITLE
Bug/signup password validation #197

### PR DESCRIPTION
**Pull request by EspressoCode**
# Related Issue
- This pr is a bug fix for:  #197  

# Proposed changes
- Change the error message for prompting user to have at list one symbol in password, now by showing a given list of symbols.
- Password match should work fine now by adding a custom validator to the form (instead of dynamically setting the validator).

# Reviewer(s)
 - @Alessandro100 @wltrf  @DPigeon

### CODE REVIEW

File: client/src/app/components/signup-form/signup-form.component.ts [hunk 1]
```diff
@@ -126,11 +131,16 @@ export class SignupFormComponent {
           return 'Email is not in the valid format.';
 
         case 'pwd,pattern':
-          return 'Password must be at least 8 characters, with one number, one letter and one symbol.';
+          return 'Password must contain at least 8 characters, including one number, one letter and one of these special characters: @$!%*#?&';
```
Comment (wltrf) [inline client/src/app/components/signup-form/signup-form.component.ts:134]: Could add: "and no space"

**Review by wltrf — CHANGES_REQUESTED** [review]

File: client/src/app/components/signup-form/signup-form.component.ts [hunk 2]
```diff
@@ -23,40 +24,46 @@ export class SignupFormComponent {
...
+        email: [
+          '',
+          Validators.compose([Validators.required, Validators.email]),
+        ],
+        pwd: [
+          '',
+          Validators.compose([
+            Validators.required,
+            Validators.pattern(
+              /^(?=.*[A-Za-z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/u
```
[the remark this answers is not in the document]
Comment (Alessandro100) [reply client/src/app/components/signup-form/signup-form.component.ts:59]: I have to agree with walter, I think our password for now should only have the validation of 8 character length, nothing more

[the remark this answers is not in the document]
Comment (wltrf) [reply client/src/app/components/signup-form/signup-form.component.ts:59]: ^ Thats a bit too extreme imo as the validation work is already done and works, the issues I stated above is 1. we need to accept more than 8 different special chars, and 2. we need to accept any character except whitespace for the overall password. The fix is quick and easy just replace the regex

[the remark this answers is not in the document]
Comment (EspressoCode) [reply client/src/app/components/signup-form/signup-form.component.ts]: I agree, sorry that was my bad, I just copied and pasted what I had for the first part of  my regex into the second part without thinking about what exactly it does. In the new commit I fixed it.

**Comment by wltrf** [comment]
Screenshot from 2020-12-10 16-52-35
@EspressoCode @Alessandro100  Why did we remove a feature that worked?

File: client/src/app/components/signup-form/signup-form.component.ts [hunk 3]
```diff
@@ -126,11 +129,17 @@ export class SignupFormComponent {
           return 'Email is not in the valid format.';
 
         case 'pwd,pattern':
-          return 'Password must be at least 8 characters, with one number, one letter and one symbol.';
+          return 'Password must contain at least 8 characters, including one number, one lowercase and one UPPERCASE letter, and no space.';
```
Comment (Alessandro100) [inline client/src/app/components/signup-form/signup-form.component.ts:132]: we can write 'uppercase' in lower case

**Review by Alessandro100 — CHANGES_REQUESTED** [review]
I made a small comment about writing 'UPPERCASE' in lowercase. I noticed that if you have an invalid password that doesn't match, it will prioritize showing the 'your password doesn't match' message. Would it be possible to change the priority to 'Your password needs xxxx' 

image

**Review by wltrf — APPROVED** [review]

**Review by Alessandro100 — APPROVED** [review]
Great fixes Hossein! Looks good to me

**Pull request merged**